### PR TITLE
[NFC] Simplify rec group initialization in wasm-type.cpp

### PR DESCRIPTION
**Pull request by tlively**
Now that we no longer support constructing basic heap types in TypeBuilder, we
can fully initialize rec groups when they are created, rather than having to
initialize them later during the build step after any basic types have been
canonicalized. Alongside that change, also simplify the process of initializing
a type builder slot to avoid completely overwriting the HeapTypeInfo in the slot
and avoid the hacky workarounds that required.

### CODE REVIEW

**Comment by tlively** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5683** <a href="https://app.graphite.dev/github/pr/WebAssembly/binaryen/5683" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/WebAssembly/binaryen/5683?utm_source=stack-comment).

File: src/wasm/wasm-type.cpp [hunk 1]
```diff
@@ -2247,14 +2248,21 @@ struct TypeBuilder::Impl {
...
-      hti.recGroup = info->recGroup;
-      *info = std::move(hti);
-      info->isTemp = true;
+      info->kind = hti.kind;
+      switch (info->kind) {
+        case HeapTypeInfo::SignatureKind:
+          info->signature = hti.signature;
+          break;
+        case HeapTypeInfo::StructKind:
+          info->struct_ = std::move(hti.struct_);
```
Comment (kripken) [inline src/wasm/wasm-type.cpp:2260]: Why is this moved but nothing else?

**Review by kripken — APPROVED** [review]
